### ISSUE TITLE
New Cool Stuff for ELKS

### ISSUE THREAD

**Issue by ghaerr** [ghaerr/elks#1366]
I'm opening this issue to talk about various projects I'm working on, most of which are running, but not fully integrated into the ELKS build yet. I'm interested to know how interested people might be in each. Here's the first three projects:
- ELKS remote console display
- D-Flat Text UI windowing toolkit
- Debug stack traces for ELKS executables

I'll write up each of these separately.

**Comment by ghaerr** [comment]
ELKS Remote Console

This little program allows one to login through a serial or telnet connection to ELKS (using a macOS or Linux terminal program, for instance), and display what is currently displayed on the ELKS console. This could be useful to see remote kernel messages, or just see exactly what is displayed in full color, using the ROM BIOS 437 character set.

For example, here's a login from macOS Terminal over emulated serial on QEMU, just after boot:
<img width="752" alt="Screen Shot 2022-07-11 at 9 08 23 PM" src="https://user-images.githubusercontent.com/11985637/178402012-10757198-e90a-49f2-bfc5-d3cb203b7c01.png">

The display is updated every second, and the program is terminated with ^C.

The `cons` program reads the PC text RAM contents directly and converts the characters into UTF-8 encoded Unicode values, along with interpreting the PC character attributes (color, reverse video, etc) and translating those into ANSI escape sequences.

The current implementation just displays the console screen, it doesn't try to insert keystrokes, etc.

**Comment by ghaerr** [comment]
D-Flat Text User Interface Library

D-Flat is an old (1995 era) text user interface that allowed overlapping window-based text user interfaces to be written on DOS. I've ported it to UNIX/Linux, and now to ELKS. It is very well written, but unfortunately kind of large, and requires medium model (large code) for compilation. The underlying API is message-based, and very similar to Win32, with SendMessage and window procedures. 

Here's a screenshot of the D-Flat `memopad` program, which allows for editing files:
<img width="832" alt="Screen Shot 2022-07-11 at 9 12 35 PM" src="https://user-images.githubusercontent.com/11985637/178402901-abff0747-653d-4519-8de4-13568e5efbe1.png">

I'm primarily working on it as a possible TUI for the Cosmopolitan project, but thought it would be interesting to get it running on ELKS.

Now to show off the `cons` program more, here's the cons program running on serial macOS Terminal, while `memopad` happens to be running on the console:
<img width="752" alt="Screen Shot 2022-07-11 at 9 13 21 PM" src="https://user-images.githubusercontent.com/11985637/178403088-78796322-72f8-4071-bd59-83758f8651c9.png">

**Comment by ghaerr** [comment]
Debug library for ELKS applications

While trying to get larger ELKS programs working, I've found it quite hard to understand what is going on when the program crashes, especially with no debugger or stack tracer. I've developed a set of routines that will allow for a running ELKS program (currently applications only, but could be fitted into the kernel itself for device drivers etc) to produce a stack backtrace, along with correct lookup of the actual C symbol names.

I've also implemented Cosmopolitan-style C function logging capability, which allows for display of every C function executed, to make it possible to visualize function call flow throughout program execution. I plan on adding nanosecond-precision timing display (on 386+ CPUs only, using the RDTSC instruction) to show function latencies as well. The idea would be to add an automatic  "--ftrace" program argument that would turn on this display when an application is compiled for debug.

I took ELKS `basic` and added the stack trace capability on every function call, here are the first few functions called, along with their arguments. The display also shows the current and max stack used by the program:
```
ELKS 0.6.0

login: root
# basic
AX=0000 BX=7460 CX=0001 DX=7464 SI=7460 DI=3448 BP=7446 SP=7422
CS:IP=5540:36f0 DS=6200 ES=6200 SS:SP=6200:7422

Level Addr    BP   DI   SI   Ret  Arg  Arg2 Arg3 Arg4
~~~~~ ~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   0: 7418 => 7446 3448 7460 36f7 dead 0000 0000 0000 0000 0000 0000 0000 0000 0000 print_stack (364e,83)
   1: 7446 => 744e 2c60 7460 3448 745c 2c60 2cb0 0019 0001 7460 7464 0000 0001 7472 __cyg_profile_func_enter+35 (36f7,83)
   2: 744e => 745c 2c60 2cb0 0019 0001 7460 7464 0000 0001 7472 0000 7478 7482 748d main+b (3448,83)
   3: 745c => 0000 0001 7472 0000 7478 7482 748d 7497 74a1 74b2 0000 6162 6973 0063 _start+19 (0019,81)
   4: 0000 => 0000 0000 5825 0900 2800 2900 2b00 2d00 2a00 2f00 3d00 3e00 3c00 3c00 _start+1 (0001,81)
>main, from _start+19, stack 0/0
AX=0019 BX=0000 CX=5401 DX=0000 SI=7460 DI=2ce5 BP=7414 SP=73f0
CS:IP=5540:36f0 DS=6200 ES=6200 SS:SP=6200:73f0

Level Addr    BP   DI   SI   Ret  Arg  Arg2 Arg3 Arg4
~~~~~ ~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   0: 73e6 => 7414 2ce5 7460 36f7 dead 00a1 0019 73fe 73fe 354e 7408 1742 1c9c 5014 print_stack (364e,83)
   1: 7414 => 7442 0000 7460 2ce5 0000 3448 0000 0000 0000 735f 6174 7472 312b 0000 __cyg_profile_func_enter+35 (36f7,83)
   2: 7442 => 744e 0000 7460 345b 0000 46b8 745c 2c60 2cb0 0019 0001 7460 7464 0000 tty_isig+b (2ce5,83)
   3: 744e => 745c 2c60 2cb0 0019 0001 7460 7464 0000 0001 7472 0000 7478 7482 748d main+1e (345b,83)
   4: 745c => 0000 0001 7472 0000 7478 7482 748d 7497 74a1 74b2 0000 6162 6973 0063 _start+19 (0019,81)
   5: 0000 => 0000 0000 5825 0900 2800 2900 2b00 2d00 2a00 2f00 3d00 3e00 3c00 3c00 _start+1 (0001,81)
|>tty_isig, from main+1e, stack 25/25
```
I am also thinking of a use where a function backtrace would be displayed whenever ^C (SIGINT) is typed, to show what the program is doing at the time of interrupt. This would be very useful for debugging hung programs.

The instrumentation facilities used to implement function call tracing require a patch to the ELKS compiler toolchain, and the stack trace capabilities require some special link handling which adds debug symbol information into the current executable, both of which are rather complicated at the moment.

In addition, I have some library code that can interpret memory bytes and display as 8086 disassembly, running directly on ELKS. This could be used within an application, or added to the `hd` or a new `disasm` or monitor/debug program, for better visibility of very low-level stuff while running on ELKS. The `hd` program is going to be enhanced to display codepage 437 (ROM BIOS) characters on a terminal display, to better see binary values as part of a hex dump.

**Comment by Mellvik** [comment]
Very interesting indeed, @ghaerr.
I strongly support the ^C stack trace idea, and by extension - the ability to send a process a signal to have it display a 'current' stack trace to the console.
Which in turn would make the cons program (neat idea!) immediately invaluable.

Apropos nanosecond resolution - don't you see a challenge in the way the output would affect the timing?

Disasm: Sounds very useful, but I didn't understand how it may be used. Are you thinking like 'hd -disasm /dev/mem' (actually some part of /dev/mem)?

Thank you - this is great work!

--M

**Comment by tyama501** [comment]
Hello @ghaerr ,

Thank you for very cool stuffs!
I started to port nano-X to pc-98 but it seems harder than I thought so it would be nice to have the debug library.

D-Flat is also interesting to see applications like DOS edit on ELKS but I'm not sure it will work for PC-98 too.

**Comment by ghaerr** [comment]
Hello @tyama501,

> I started to port nano-X to pc-98 but it seems harder than I thought

It may be easiest to just copy nano-X/drivers/scr_bios.c to scr_pc98.c, and plan on having all graphics draw code in that single file. Rename all the functions from VGA_ to pc98_, and then start with just writing the pc98_open() and pc98_close() routines which should switch the screen to and from graphics mode, using the same code from BASIC. 

You should then be able to write the very basic pc98_drawpixel, pc98_drawhline, pc98_drawvline  and pc98_fillrect functions, which do not depend on an internal graphics format. (HAVEBLIT should be set to 0).

After that, in order to support BLIT, we would need to make a decision as to how nano-X should store the graphics data internally, which is set with the psd->app and other statements in pc98_open().

If you'd like to share what you have, I'm happy to help you get the screen driver working. The mouse driver, but default uses the serial port, which isn't yet implemented for PC-98, so you might want to compile in drivers/mou_null.c.

> D-Flat is also interesting to see applications like DOS edit on ELKS but I'm not sure it will work for PC-98 too.

What is the character set for byte values 128 - 255 in PC-98 character ROM? These won't be codepage 437, so that could require low-level changes to D-Flat for running on console. However, I am thinking having the library code automatically switch to using ANSI ESC sequences when running on telnet or serial, and having the CP 437 codes converted to unicode for that case (the same was `cons` works, basically).

Thank you!

**Comment by ghaerr** [comment]
Hello @Mellvik,

> Apropos nanosecond resolution - don't you see a challenge in the way the output would affect the timing?

Yes, the trace and timer code will definitely add overhead to all function tracing, but the idea is it will be a fixed amount, and will adjust all timings by that same fixed amount, so bottleneck latencies can still be easily spotted. The routine could calculate its own overhead and subtract that from the displayed time for better accuracy (I've added that thought to my list).

> Disasm: Sounds very useful, but I didn't understand how it may be used.

I think the most useful case would be some kind of debug/monitor that would trap to a prompt (interrupting the running program from a SIGINT), then producing a backtrace and allow the user to dump register or specify a seg:off memory address to start disassembly from. As a library routine, it would also be useful to write programs to disassemble various files at times. Another use could be instruction tracing, by allowing a single instruction to execute on real hardware, followed by an INT 3 back to the monitor. This would work because the disassembler has the ability to compute how long a given instruction sequence is.

I'm not quite sure where to put all this - I suppose just adding the the C library would be OK, except these are all non-standard routines. So perhaps a separate library... not sure yet.

I'm also working on being able to compile and link programs outside of the ELKS source tree - this is too hard currently, as there's no easy way to get all the compiler options and C library. Once this is done it will be lots easier to contribute programs to ELKS, without having to become intimately familiar with the kernel and application source trees.

Thank you!

**Comment by Mellvik** [comment]
Thank you @ghaerr -
> Disasm: Sounds very useful, but I didn't understand how it may be used.
> 
> I think the most useful case would be some kind of debug/monitor that would trap to a prompt (interrupting the running program from a SIGINT), then producing a backtrace and allow the user to dump register or specify a seg:off memory address to start disassembly from. As a library routine, it would also be useful to write programs to disassemble various files at times. Another use could be instruction tracing, by allowing a single instruction to execute on real hardware, followed by an INT 3 back to the monitor. This would work because the disassembler has the ability to compute how long a given instruction sequence is.
> 
Wow, that's ambitious indeed!! I'm rather looking forward to it - exciting! 
> I'm not quite sure where to put all this - I suppose just adding the the C library would be OK, except these are all non-standard routines. So perhaps a separate library... not sure yet.
> 
My 2cents - separate library.
> I'm also working on being able to compile and link programs outside of the ELKS source tree - this is too hard currently, as there's no easy way to get all the compiler options and C library. Once this is done it will be lots easier to contribute programs to ELKS, without having to become intimately familiar with the kernel and application source trees.
> 
Agreed, and again very useful indeed!


ELKS rocks


-M

**Comment by tyama501** [comment]
Hello @ghaerr 
 
> It may be easiest to just copy nano-X/drivers/scr_bios.c to scr_pc98.c, and plan on having all graphics draw code in that single file. 

I have copied vgaplan4.c to vgaplan4_pc98.c and I have been modifiying ega_drawpixel to draw directly to 0xA800, 0xB000, 0xB800, 0xE000 VRAM after changing to the graphic mode from basic but your idea sounds easier for the first step.


> the serial port, which isn't yet implemented for PC-98, so you might want to compile in drivers/mou_null.c.

I have noticed this, and I commented out GdOpenMouse in srvmain.c for now.
OK, I will use mou_null instead.

For PC-98 there is no serial mouse, and the mouse is attached to 8255 I/O port.
So, maybe we need to write the driver with polling or interrupt after the screen driver worked.

> > D-Flat is also interesting to see applications like DOS edit on ELKS but I'm not sure it will work for PC-98 too.
> 
> What is the character set for byte values 128 - 255 in PC-98 character ROM? 

I poked to 0 - 255 directly to text vram with basic :)
It seems there are single byte katakana-characters and some others.

![pokeb_0to255_text](https://user-images.githubusercontent.com/61556504/178785134-64be7450-087f-41ba-b3ea-0735a9d14ffe.png)

Thank you.

**Comment by ghaerr** [comment]
Hello @tyama501,

> I have copied vgaplan4.c to vgaplan4_pc98.c and I have been modifiying ega_drawpixel to draw directly to 0xA800, 0xB000, 0xB800, 0xE000 VRAM after changing to the graphic mode

I see. I suppose it depends on how close the PC-98 graphics hardware is to EGA - EGA is terrible, and requires multiple mask and plane register to be set for each pixel. If that's the case for PC-98, perhaps staying with your present approach is better. Another thought would be to only support mono (2-color) mode initially, is that possible, then possibly not having to write to each memory plane for each pixel.

After getting running, you will want to "define NDEBUG" to remove the assert() code, as it slows things down considerably.

> For PC-98 there is no serial mouse, and the mouse is attached to 8255 I/O port.
So, maybe we need to write the driver with polling or interrupt after the screen driver worked.

When you use mou_null, this should happen automatically. Afterwards, a PC-98 mouse driver can be written.

In general, nano-X busy-loops by polling, so this shouldn't be a problem unless the polling itself is quite slow.

> It seems there are single byte katakana-characters and some others.

I can see it's definitely not codepage 437, but does include the box line draw characters. There doesn't seem to be the 50% gray shaded character that is used for the background of D-Flat windows, but that isn't a big problem, a space can be used. 

Should we get more interested in actually running D-Flat on PC-98, we can develop a translation table for the different code page, both for D-Flat use, as well as for the `cons` unicode translation. 

Thank you!

**Comment by toncho11** [comment]
Hi @ghaerr,

I am for D-Flat project - it seems it can produce a nice DOS editor and some other applications as well. We can have a developer section one day: "Programming for ELSK" using:
- pure C code
- using X-nano
- using D-Flat TUI
- BASIC

**Comment by ghaerr** [comment]
D-Flat and its text-mode user interface library demo is now running on ELKS connecting from serial or telnet terminal sessions. Here's a video showing it in operation. In this particular case, all I/O is based on ANSI terminal escape sequences, so the TUI display is available to many remote systems, using any modern terminal emulator, such as macOS Terminal or Linux xterm-256.

It also runs on the console, but this requires more setup as a mouse driver that spits out ANSI sequences needs to be installed, unless the program can be entirely operated from the keyboard.

The video is too big for Github upload, [here's a link to my dropbox for viewing](https://www.dropbox.com/s/qlssrndb007vjwa/D-Flat%20ELKS%20Demo.mov?dl=0).

**Comment by toncho11** [comment]
Nice job @ghaerr !!!
So it can run natively (without any remote sessions) on ELKS, but one needs to use the keyboard shortcuts.

> D-Flat and its text-mode user interface library demo is now running on ELKS connecting from serial or telnet terminal sessions. Here's a video showing it in operation. In this particular case, all I/O is based on ANSI terminal escape sequences, so the TUI display is available to many remote systems, using any modern terminal emulator, such as macOS Terminal or Linux xterm-256.
> 
> It also runs on the console, but this requires more setup as a mouse driver that spits out ANSI sequences needs to be installed, unless the program can be entirely operated from the keyboard.
> 
> The video is too big for Github upload, [here's a link to my dropbox for viewing](https://www.dropbox.com/s/qlssrndb007vjwa/D-Flat%20ELKS%20Demo.mov?dl=0).

**Comment by Mellvik** [comment]
Wow @ghaerr - this is impressive.

Apropos mouse: Can the interface be operated w/o a pointing device?


-M

**Comment by ghaerr** [comment]
> Can the interface be operated w/o a pointing device?

Yes, I'm sure it could, but not certain as to what has been coded in D-Flat. D-Flat is a somewhat large text user interface (TUI) library, by ELKS' standards. The `memopad` demo, which includes all the TUI controls (scrollbars, listboxes, popups, comboboxes, edit boxes, etc) is 110k - huge. D-Flat is IBM SAA/CUA compliant as of their [1987 standard](https://en.wikipedia.org/wiki/IBM_Common_User_Access).

However, the real purpose of the work is mostly R&D, and under the hood portions of the following are already implemented, or being considered:
- Working medium model ELKS applications (a toolchain bug was [recently fixed](https://github.com/tkchia/gcc-ia16/issues/117) getting memopad to load properly).
- Writing TUI applications that compile and work almost identically between macOS, Linux and ELKS.
- Unicode support, which allow ELKS to run applications that can display unicode (using UTF-8) on serial or telnet connections. In particular, the ability to display BIOS character ROM codepage 437 characters external to the console is a big win.
- Possible solutions to long-standing character display problems such as #1175 because the PC ROM character set isn't the same across machines internationally.
- Enhancement of incoming and outgoing ANSI escape sequence support for ELKS console and terminal sessions, including accepting a wider ranger sequences on direct console for UNIX application compatibility, as well as ANSI mouse and window sizing support being put in library routines.
- A possible ability for an application to know whether its running on direct console or serial/telnet terminal, and to automatically perform character conversions via Unicode and/or ANSI sequences for visual compatibility. (For instance, the `cons` console display program, and planned `hd` display enhancements to see the full CP437 character set).
- Possible support for arbitrary unicode on non-console applications. This is important, as the rest of the world is now running on unicode, and ANSI sequences allow unicode display on all modern terminal emulators.
- Development of multiple layers of TUI support, from automatic mouse/keyboard support without termcap or ncurses (using ANSI/xterm standard) for building portable TUI applications, the ability to write characters/attributes to the screen directly for speed, all the way to heavier-weight libraries like D-Flat.
- Test programs that allow for easier unit-testing of various library components.
- Testing of the kernel and C library implementation by using larger pieces of software (more incompatibilities have been found).

All this said, the fact that ELKS is basically running almost all of D-Flat identically to its UNIX counterpart is a testament to how far ELKS has come in the last few years - its pretty amazing our 16-bit system can keep up and run more modern UNIX 32- and 64-bit programs, thanks to everyone's work :)

**Comment by toncho11** [comment]
@ghaerr Are you planning on submitting D-Flat as part of the bin folder? Maybe only for the HDD images?

**Comment by toncho11** [comment]
@ghaerr  Actually as a binary release it will be only memopad?

**Comment by ghaerr** [comment]
Hello @toncho11,

My plan would be to consider adding the D-Flat source code as a text user interface library, along with a number of other helpful library routines, somewhere in the ELKS source tree. The library would be automatically built so that others could build their own TUI applications. The D-Flat library includes only a single demo `memopad` that displays all of its capabilities which would also be copied to the /bin directory for any images 2.88M and larger.

The real problem with most TUI and GUI libraries is that it isn't trivial to build an application using them. Frankly, it isn't real clear to me what anyone might build with a D-Flat TUI. (Given that the `kilo` and `nano` editors don't work well, we renamed the MINIX screen `mined` to `edit`, and it seems to work quite well, so we don't need a screen editor in D-Flat). Even though D-Flat is well thought-out and well-written, it is large by ELKS standards, and programs produced using the library won't fit well on floppy disks with the rest of the standard distribution.

Thus, D-Flat itself is mostly a research project, which (described above) is helping to illuminate some things that might want to be considered next for development. That said, it's opening the doors for more discussion on what we should next add to ELKS. Thank you for your comments!

**Comment by Mellvik** [comment]
Again - an impressive undertaking @ghaerr, and - as you point out - an important demo of how far ELKS has come. It is indeed impressive what can be tweaked out of limited resources with smartness, patience and attention to detail (read: avoid the bloat risk). I got a vivid demonstration of this the other day  when I had to boot W95 on a 386sx in order to check out some PnP issues. Even at 40MHz and w/16MB mem W95 is totally useless while the resources are overkill for ELKS (feels like QEMU). Not a fair comparison, but a reminder. ELKS is (or can be) one floppy, W95 is 31 floppies to get started :-).
> However, the real purpose of the work is mostly R&D, and under the hood portions of the following are already implemented, or being considered:
> 
An important angle - and ambition: Create a platform for new ideas, experimentation rather than being a small scale copy of larger siblings. Or simply creating a tool for some special purpose. My 2C along that angle of thinking:
> Working medium model ELKS applications (a toolchain bug was recently fixed getting memopad to load properly).
Important, resource utilization (btw - your recent work on debugging infrastructure falls into this category).
> Writing TUI applications that compile and work almost identically between macOS, Linux and ELKS.
> Unicode support, which allow ELKS to run applications that can display unicode (using UTF-8) on serial or telnet connections. In particular, the ability to display BIOS character ROM codepage 437 characters external to the console is a big win.
Agreed - UTF-8 is an important step.
> Possible solutions to long-standing character display problems such as #1175 because the PC ROM character set isn't the same across machines internationally.
> Enhancement of incoming and outgoing ANSI escape sequence support for ELKS console and terminal sessions, including accepting a wider ranger sequences on direct console for UNIX application compatibility, as well as ANSI mouse and window sizing support being put in library routines.
Important - and challenging. Reliably reading multibyte ANSI sequences is touch on both serial and telnet, even on Linux and RaspberryPis with vastly more resources than ELKS. And frequently evidenced in ELKS and vi today - on serial.
> A possible ability for an application to know whether its running on direct console or serial/telnet terminal, and to automatically perform character conversions via Unicode and/or ANSI sequences for visual compatibility. (For instance, the cons console display program, and planned hd display enhancements to see the full CP437 character set).
> Possible support for arbitrary unicode on non-console applications. This is important, as the rest of the world is now running on unicode, and ANSI sequences allow unicode display on all modern terminal emulators.
Agreed, immediately very useful.
> Development of multiple layers of TUI support, from automatic mouse/keyboard support without termcap or ncurses (using ANSI/xterm standard) for building portable TUI applications, the ability to write characters/attributes to the screen directly for speed, all the way to heavier-weight libraries like D-Flat.
> Test programs that allow for easier unit-testing of various library components.
> Testing of the kernel and C library implementation by using larger pieces of software (more incompatibilities have been found).
Indeed. Beating up the system in practical usage is when the 'deep' bugs pop up. The buffer bug I'm on right now being a great case in point.
> All this said, the fact that ELKS is basically running almost all of D-Flat identically to its UNIX counterpart is a testament to how far ELKS has come in the last few years - its pretty amazing our 16-bit system can keep up and run more modern UNIX 32- and 64-bit programs, thanks to everyone's work :)
> 
A final point: We (the world) are heading into rough waters these days, and will need to become more resource conscious. To me ELKS is part of that. Industrial control systems in the thousands that were deemed obsolete and useless years ago can do serious work with better platforms - like ELKS. The 386sx system mentioned above is a great example: a 20+ years old SBC perfectly useable for a lot of purposes with the 'right' ('unbloated') software. Taking that angle, ELKS development or more than a hobby, it's a contribution to sustainability. 

Thanks again, this is great work.

--M

**Comment by ATroubleshooter** [comment]
> 
> Thus, D-Flat itself is mostly a research project, which (described above) is helping to illuminate some things that might want to be considered next for development. That said, it's opening the doors for more discussion on what we should next add to ELKS. Thank you for your comments!

Considering D-Flat, in regards to as what should be added next - probably, a simplistic lightweight twin-panel file manager, based on D-Flat?

**Comment by ghaerr** [comment]
> Considering D-Flat, in regards to as what should be added next - probably, a simplistic lightweight twin-panel file manager, based on D-Flat?

That's a nice idea, but potentially a lot of work to (re)write a nice file manager. What would help would be something like a 16-bit version of Midnight Commander, which was discussed a bit in #1117, but apparently not available. With the D-Flat lower level ANSI/xterm input/output, we could eliminate the need for the huge and clunky ncurses library, and only use xterm/ANSI sequences, which would display properly on ELKS and most all Linux/macOS systems. That is, an older ncurses-based file manager could be slimmed down and possibly more easily ported by replacing ncurses with the D-Flat lower level.

Does anyone have any ideas about an older, small file manager that might be portable?

**Comment by ghaerr** [comment]
Looking around for other "small" text-based file managers, I finally found two, both based on ncurses: `nnn` (https://github.com/jarun/nnn) which is highly supported and does lots of stuff. It purports to running well on embedded systems. Here's a screenshot:
<img width="752" alt="nnn screenshot" src="https://user-images.githubusercontent.com/11985637/180275981-e40357a6-f673-43e1-9c91-48fde8ecf406.png">

However, in reading a comparison article, `nnn` apparently uses far less memory than other Linux-based file managers, "only" 3.5MB of RAM!!! (compared to 9.8MB for Midnight Commander, and 51MB for the largest):
![nnn-Memory](https://user-images.githubusercontent.com/11985637/180276263-42be001d-2a95-4cce-9882-747f97d1507d.png) The `nnn` executable on macOS is 131K with no options enabled.

I guess RAM is cheap.

Looking further, I find `noice` (https://github.com/RichardHyde/noice), which it turns out `nnn` is forked from. It compiled quickly on macOS, and is a 29K executable. Its development stopped in 2015, but it looks very similar, just doesn't use any color display:
<img width="752" alt="noice screenshot" src="https://user-images.githubusercontent.com/11985637/180276803-2d70d682-7faa-4e9e-ae31-a0c2dd1f96cc.png">

The nnn source is 8800 lines, and noice is 1000. Looking at noice, it seems it could be ported to ELKS quickly, with an initial version using just the arrow keys (and/or mouse on a remote terminal), using the low-level D-Flat framework I just developed. Later, a version could be possibly inserted into a D-Flat frame to look cooler. I haven't looked at the memory usage yet for noice. I may take a pass at porting noice to ELKS if people are interested.

**Comment by ghaerr** [comment]
I looked further into `noice` and found the original project (https://git.2f30.org/noice), which was last updated in 2019. It is a quite small and nifty, portable file manager.

I ported this to ELKS as our new `fm` file manager, and wrote an ncurses library emulation layer for it, which allowed `fm` to run directly on the console or from a serial/telnet terminal. The file manager is only 13K bytes, so it should run well on floppies.

Here's a screen recording of it in operation. I plan on porting the lower layer to be able to use the mouse or function keys on the console or serial/telnet terminal, and then ultimately should be pretty straightforward to put into a D-Flat window. Pretty cool!

https://user-images.githubusercontent.com/11985637/180363966-87eef0b8-ae6e-4f26-a284-e87f93935bce.mov

Thanks for the idea @ATroubleshooter :)

**Comment by Mellvik** [comment]
Indeed @ghaerr - this IS cool!

> 22. jul. 2022 kl. 06:49 skrev Gregory Haerr ***@***.***>:
> 
> Here's a screen recording of it in operation. I plan on porting the lower layer to be able to use the mouse or function keys on the console or serial/telnet terminal, and then ultimately should be pretty straightforward to put into a D-Flat window. Pretty cool!
> 
> 


-M

**Comment by ATroubleshooter** [comment]
Great job @ghaerr, keep it up!

**Comment by ghaerr** [comment]
Update on new File Manager for ELKS and UNIX:

Significant progress has been made, the the project now compiles and runs identically on ELKS console, ELKS terminals, as well as UNIX (tested currently only on macOS Terminal). It seems to work pretty well, and for now have created its own [project](https://github.com/ghaerr/fm), before integrating directly into ELKS. 

Here's a screenshot running on macOS (which pretty much looks identical to the ELKS terminal version):
<img width="752" alt="FM terminal" src="https://user-images.githubusercontent.com/11985637/180909279-32381dd1-95ac-491f-a47d-d58fb0811521.png">

The file manager has lots of features for its small size, but here are the basic keystrokes:
```
RightArrow/Enter - Enter directory or display file
LeftArrow/Backspace - Back up one directory
Up/DownArrow or mouse scroll wheel - move cursor
a-z - goto file matching letter pressed
Z or ^C - quit
D - directories displayed first
T - sort by time
S - sort by size
C - change directory
E - edit file
. - display hidden files
~ - home directory
? - help
```
The program automatically decompresses .Z files for display, and will run `man` on man page files in compressed or normal format. `more` is used to display files.

After cloning the source tree, type `make` for the Linux version, or `make -f Makefile.elks` for the ELKS version, then type `./fm`.

I've worked hard to make the program as small as possible, which is pretty small since ncurses and regex libraries were removed. `fm` requires and ANSI (or xterm-256) terminal to run.

The ELKS version is 20k and UNIX 40k. Let me know what you think. At some later point I may wrap the output in D-Flat to look more like Midnight Commander.

Enjoy!

**Comment by toncho11** [comment]
Great job @ghaerr !

**Comment by toncho11** [comment]
I see edit is possible.
Here are suggestions for future versions:
- configure editor (which is the default now?) 
- copy files with progress bar
- delete and move both files and directories

Looking back I like Norton Commander, but PC Tools is the one that impressed me the most back then. PCtools is not double screen. A new window pops up and you can select directory tree or drive (I think).

**Comment by ghaerr** [comment]
> configure editor (which is the default now?)

`vi` is the default, but any editor can be specified using the EDITOR= environment variable.

> copy files with progress bar
> delete and move both files and directories

Good ideas, but `fm` is currently just single-paned. I'll wait and see whether people actually use it before spending too much time on it. With the under-the-hood work on the keyboard and mouse input for D-flat, it was really good timing to work on another program that heavily used the mouse and arrow keys. D-flat doesn't yet support the mouse wheel, and I'm going to jump back into that for bit and try to get a version others can play with on ELKS or UNIX.

**Comment by ghaerr** [comment]
Update on D-Flat for ELKS and UNIX:

A [separate project](https://github.com/ghaerr/dflat) has been created for D-Flat, where it can be compiled for macOS, Linux or ELKS. Things have been progressing nicely, and it now automatically sizes to the open Terminal on startup, just like the `fm` File Manager. Both projects share the same new "unikey" ANSI keyboard/mouse library, which also supports unicode character conversions (currently only for code page CP437). The screen updates are FAST.

Here's D-Flat's `memopad` running on macOS. The display looks identical when run from ELKS serial or telnet:
<img width="708" alt="D-Flat_Text_User_Interface_on_UNIX" src="https://user-images.githubusercontent.com/11985637/181430293-1da5b15f-8350-4c89-9c3e-adac780cbb55.png">

The new unikey library supports ANSI mouse sequences from macOS and Linux terminal programs emulating xterm-256, and I plan on adding the mouse sequence support to ELKS running QEMU over an emulated serial port. That would allow for operation of QEMU console programs using the mouse from your host computer or trackpad to operate it. On real hardware, the serial mouse is already supported, but not yet built into the unikey library, which would allow for a new class of programs to be developed for ELKS - text based user interfaces which could be sped up greatly just by using the mouse wheel for fast scrolling, for instance.

If you have time, take a look at the project and compile it. Because both D-Flat and File Manager are being developed for both UNIX and ELKS, I'm thinking that adding them as a Git submodule (i.e. a repo within the ELKS repo) in the future, which would allow for automatic compilation and inclusion on the ELKS disk images.

**Comment by tyama501** [comment]
Hello @ghaerr ,

Nice! Yes I agee adding other projects as submodules to expand ELKS world.
It would be nice if more developers contribute to ELKS through those projects.

Thank you.

**Comment by ghaerr** [comment]
With the console fixes in PR #1388, ELKS can now simulate the Matrix! :)

https://user-images.githubusercontent.com/11985637/182286487-20db97ca-94bb-4e60-b168-7c109295c9ea.mov

**Comment by toncho11** [comment]
Impressive @ghaerr :)

**Comment by ATroubleshooter** [comment]
> Update on new File Manager for ELKS and UNIX:
> 
> Significant progress has been made, the the project now compiles and runs identically on ELKS console, ELKS terminals, as well as UNIX (tested currently only on macOS Terminal). It seems to work pretty well, and for now have created its own [project](https://github.com/ghaerr/fm), before integrating directly into ELKS.
> 
> Here's a screenshot running on macOS (which pretty much looks identical to the ELKS terminal version): <img alt="FM terminal" width="752" src="https://user-images.githubusercontent.com/11985637/180909279-32381dd1-95ac-491f-a47d-d58fb0811521.png">
> 
> The file manager has lots of features for its small size, but here are the basic keystrokes:
> 
> ```
> RightArrow/Enter - Enter directory or display file
> LeftArrow/Backspace - Back up one directory
> Up/DownArrow or mouse scroll wheel - move cursor
> a-z - goto file matching letter pressed
> Z or ^C - quit
> D - directories displayed first
> T - sort by time
> S - sort by size
> C - change directory
> E - edit file
> . - display hidden files
> ~ - home directory
> ? - help
> ```
> 
> The program automatically decompresses .Z files for display, and will run `man` on man page files in compressed or normal format. `more` is used to display files.
> 
> After cloning the source tree, type `make` for the Linux version, or `make -f Makefile.elks` for the ELKS version, then type `./fm`.
> 
> I've worked hard to make the program as small as possible, which is pretty small since ncurses and regex libraries were removed. `fm` requires and ANSI (or xterm-256) terminal to run.
> 
> The ELKS version is 20k and UNIX 40k. Let me know what you think. At some later point I may wrap the output in D-Flat to look more like Midnight Commander.
> 
> Enjoy!

Thank you, @ghaerr.

I tried it on ELKS and it worked! And as for the MC-like appearance - I would really like to have it on ELKS. D-Flat could indeed make the difference. For example, this is how DOS Navigator looks, it's using "Turbo Vision" style interface too:
https://youtu.be/Fmauh_ERaag?t=325

It's got its own windowed viewer, editor with clipboard buffer support etc...

**Comment by Mellvik** [comment]
Thank you @ghaerr - 
again - this is cool. I just had to check it out right away - and sure enough, works almost out of the box (the Makefile.elks needs to be edited to set TOPDIR).

How do I activate the menu/instruction line on the bottom?
Like kilo it does a repaint for every keystroke, which is bearable at 57600bps but still very visible. IIRC that's a hard one to fix, right?
(BTW I haven't merged recent ELKS updates, that may be part of it).


—M

**Comment by ghaerr** [comment]
Hello @ATroubleshooter,

> And as for the MC-like appearance - I would really like to have it on ELKS. D-Flat could indeed make the difference.

Thanks for trying out `fm` :) I agree, I plan on enhancing it to look like MC, but wanted to see whether users find it useful in the first place. It's on my list to improve, and the D-Flat library could be a good choice for that, although it has the problem of being kind of big.

> How do I activate the menu/instruction line on the bottom?

@Mellvik, use `?` for that. I plan on adding a full screen of help, but haven't to to that yet either.

> Like kilo it does a repaint for every keystroke, which is bearable at 57600bps but still very visible. IIRC that's a hard one to fix, right?

It doesn't repaint on every keystroke, but does repaint the entire screen during each function. I'm interested in your continued comments on usability - I was thinking this could possibly be upgraded to be the front-end of a way to upgrade various portions of ELKS or to at least run shell scripts quickly.

Thank you!

**Comment by ghaerr** [comment]
I've been working for a while on a visual emulator/debugger for ELKS binaries and the kernel itself. It's finally here and ready to play with. It's called [blink16](https://github.com/ghaerr/blink16) and is based off the incredible [Blink Project](https://github.com/jart/blink). 

It is still under construction but allows for 8086-only ELKS and DOS binaries to be loaded and the programs to be emulated on your desktop, translating system calls and displaying console I/O during execution.

A separate mode of running `blink16` allows 8086 PC emulation, where an ELKS image, with symbol table, is booted just as on a PC, and the entire internal operation of the ELKS kernel can be inspected and traced. Download the source and run `make elks` to demo a prebuilt ELKS image booting. (More instructions on the [site](https://github.com/ghaerr/blink16)).

Here's a screenshot of ELKS running 'banner ELKS', showing `banner` executing the `write` system call:

<img width="1403" alt="blink16-banner" src="https://user-images.githubusercontent.com/11985637/221447514-6a813f98-4716-47a5-a526-b49e9599a120.png">

**Comment by toncho11** [comment]
Is there video emulation? Can I run nxtetris on a contemporary x64 Linux machine by just pointing to the executable: `./blink16 nxtetris ELKS`.

**Comment by tyama501** [comment]
Hello @ghaerr ,

Nice!
Can I use it to debug elksmoria if I put files in the ibm ELKS image?

**Comment by ghaerr** [comment]
Hello @toncho11,

> Is there video emulation? Can I run nxtetris on a contemporary x64 Linux machine by just pointing to the executable

There is currently MDA and CGA emulation, but only for their text modes. It doesn't (yet) emulate EGA/VGA, that'd be probably a lot of work. We've got a lot of work to do adding in all the ELKS system calls as well (e.g. read/write/close are in, many others still missing).

> ./blink16 nxtetris ELKS.

The "ELKS:" at the end of my sample `banner ELKS` is for arguments to banner. So you'd only have to give the basic executable name, not followed by ELKS. `blink16` figures its a DOS executable if .exe extension if found, a boot block if .bin or .img is seen, otherwise ELKS executable.

I haven't tried running many programs yet, but it should run them until a missing system call is attempted. Any graphics output would just be ignored at this point, so you won't see anything Nano-X programs.

We've got the new ELKS symbol table format all working - all that's left is to automatically build all programs with a symbol table, which is coming shortly, but will (temporarily at least) increase the floppy space usage. The big benefit is that anyone can then execute a program and see its execution internally using the source code symbols. This will be a great way to get more comfortable with the internals of how any ELKS executable operates.

Thank you!

**Comment by ghaerr** [comment]
Hello @tyama501,

> Can I use it to debug elksmoria if I put files in the ibm ELKS image?

Haha, I knew you were going to ask that!! The answer is: almost. There isn't quite support working for medium model programs as far as symbol table display, although it ought to be able to emulate the program. However, emulating without a symbol table for mnemonic display makes it hard to understand where you are in the program. There is also the problem that there are many system calls missing. 

In short - it's coming.

Thank you!

**Comment by tyama501** [comment]
> In short - it's coming.

Great, thank you!

**Comment by tyama501** [comment]
I found this tweet about blink16 and ELKS:)

https://twitter.com/JustineTunney/status/1629671904696205312

**Comment by tyama501** [comment]
I could build with my FreeBSD Clang with this.

cc -DBLINK16=1 -I.. -Os -o $@ $^ -lz -DNOJIT=1 -lm

![image](https://user-images.githubusercontent.com/61556504/222215377-f14dbffe-b1b1-4324-a8f7-af0a8ccd8fbb.png)

**Comment by ghaerr** [comment]
> I could build with my FreeBSD Clang with this.
> cc -DBLINK16=1 -I.. -Os -o @^ -lz -DNOJIT=1 -lm

Wow @tyama501, nice! I hadn't realized that -DNOJIT=1 is required, I'll add that to blink16/Makefile.

Also I noticed that your terminal is running in "dark mode", I hadn't tested that as macOS Terminal runs black on white background. That looks nice. Can you expand the right side of the terminal even more? That will allow for the instruction bytes to be displayed left of the instructions themselves. I am trying to determine a way to perhaps make the memory pane views a little smaller.

Now that you're running ELKS executables, feel free to open an issue (on ghaerr/blink16) for any problems. I haven't yet tried running any medium model programs, and there are definitely issues with the symbol display, which are being worked on now. You will also find that we need quite a few more system calls to be emulated (see syscall-elks.c) in order to be useful.

We can also make changes for PC-98 specific issues if required. I think this tool will be very helpful to better understand what ELKS applications and the kernel are doing.

Thank you!

**Comment by tyama501** [comment]
Hello @ghaerr ,

Well, actually I don't know disabling jit function is ok or not.
In my environment, one of the atomic_ function in the HAVEJIT got error claiming address is const, so I just disabled it. 

Thank you.

**Comment by ghaerr** [comment]
> Well, actually I don't know disabling jit function is ok or not.

Its ok to disable, as blink16 replaced the blink VM with a much smaller 8086-only instruction interpreter that does not use JIT. Using -DNOJIT=1 is useful for the other files and includes that might possibly be included into the blink16 project. I have added -DNOJIT=1 and -lm to blink16/Makefile and will push in the next commit.

**Comment by tyama501** [comment]
Thanks

**Comment by tyama501** [comment]
Hello @ghaerr ,

![image](https://user-images.githubusercontent.com/61556504/222518130-dda7bf9c-e1dc-4387-bdb1-39517cd56c6f.png)

I deleted some files from fd1440.img to get space and put the elksmoria files.
I am trying to run it but it seems that keys are not working or too slow to enter while running with "c" command.

Is there way to speed up?

Thank you.

**Comment by ghaerr** [comment]
> I am trying to run it but it seems that keys are not working or too slow to enter while running with "c" command.
> Is there way to speed up?

For the above screenshot emulating the ELKS kernel, you can use the capital C) command, which executes instructions but does not show the instruction and memory tracing. This command still has some issues when a program is trying to read the console. For that, you can use the capital D)isplay command, which just shows console output. A ^C brings you back from that screen to the main emulation screen. I just added stack backtrace frames support, so pull that down, the output is helpful. Shortly, I will push a commit with P)rofiling support, which is cool because it shows the functions which are taking up the most time. You can press ? to get a help screen.

> I deleted some files from fd1440.img to get space and put the elksmoria files.

I'm not sure whether elksmoria will run or not, I have not tested any medium model binaries. For sure the symbol lookup won't be working yet. We also have the problem of not supporting many ELKS kernel system calls yet in application emulation mode.

(Note that blink16 runs in one of two "modes" - an "application emulation mode", that of emulating an ELKS or DOS application, and separately, a "system" mode, that of emulating an IBM PC bare machine, which is used to boot .bin sector games or .img files like the ELKS kernel itself. When blink16 is running in "system" mode (.img or .bin files), it only shows the symbols for the kernel itself, not for any running applications. Moreover, it might be hard to debug the application that way, since the timer interrupt is causing ELKS to switch between applications. The "application mode" is active when a .exe file is specified for DOS, or for ELKS when neither .bin or .img are specified.)

Another way to speed up display during c)ontinue mode is ^T (turbo). Typing that a few times speeds the display up. Blink16 also supports S)tep and N)ext, which execute a single instruction, with N) skipping over call instructions.

**Comment by toncho11** [comment]
blink16 works only on contemporary computers and it is 32/64 bit compiled, right? It is compiled with the standard gcc (not ELKS gcc16). For a second I had this idea where blink runs inside ELKS ... and it allows for the execution of DOS executables. @ghaerr this all came as I was reading that it "translates system calls". The idea is not completely crazy ... :). If it is going to be insanely slow then it could be used only on 386 and 486. If it saves you the job of having DOS installed and restarting, it's still worth it.

**Comment by tyama501** [comment]
Thank you @ghaerr ,

I see. Here is the screen after "D" and ./elksmori
![image](https://user-images.githubusercontent.com/61556504/222791315-487617be-e611-4133-bbec-c645ae97620c.png)

**Comment by tyama501** [comment]
cat morifile/newselks works.
The previous image I posted is trying to show this file.

![image](https://user-images.githubusercontent.com/61556504/222794873-ac1e0c72-d0fc-498f-afee-e7f9e677acae.png)

**Comment by ghaerr** [comment]
Hello @toncho11,

> blink16 works only on contemporary computers and it is 32/64 bit compiled, right? It is compiled with the standard gcc (not ELKS gcc16). 

Yes - blink16 is a 32/64 bit application that uses a large amount of memory to emulate the 8086 instruction set, as well as save instruction history for "rewinding" execution (by using the scroll wheel). This takes a lot of RAM and won't ever be able to run on ELKS, nor any PC with less than 32MB of RAM.


>  this all came as I was reading that it "translates system calls". The idea is not completely crazy ... :). 

The ELKS ("guest") system calls have to be translated to the operating system blink16 is running on ("host").

> If it is going to be insanely slow then it could be used only on 386 and 486. 

Not sure what you're referring to here, but if its blink16 execution speed, the purpose is for debugging, not just to emulate an 8086 like QEMU does. The capital C) and D) commands allow fast execution without instruction display.

> If it saves you the job of having DOS installed and restarting, it's still worth it.

The idea is to run some DOS executables, also by translating the DOS system calls to Linux/macOS/FreeBSD, etc, but blink16 doesn't try to emulate a full DOS environment, so I'm sure many programs may not fully work.

The intent is mostly to visually see a program's execution. Blink16 also supports P)rofiling, which shows which procedures are bottlenecks in execution time.

Thank you!

**Comment by ghaerr** [comment]
Hello @tyama501,

> Here is the screen after "D" and ./elksmori

I think you may be getting confused here on how to run blink16, please re-read my post above. That is, if you run:
```
./blink16 fd1440.img
```
This will execute blink16 in "system emulation mode", where the ELKS kernel (in this case) is booted from the .img file. Trying to run `./elksmori` from *within* system emulation mode won't work well - as blink16 is primarily emulating an 8086 system, rather than trying to run a particular ELKS program. In order to directly debug elksmori, you would want to type something like:
```
./blink16 elksmori
```
This would load `elksmori` directly into blink16 memory, then attempt running it and translate the ELKS application system calls to Linux/macOS/FreeBSD, etc. However, this will not yet work, as the blink16 loader won't load medium-model ELKS applications (yet). So you should test with another program. I have included `banner` and `echo`, built with -maout-symtab, which includes the symbol table, which really helps display.

The reason why, in the first case, ELKS won't run `./elksmori` is that the `fd1440.img` image has on it a very specially-prepared ELKS kernel which is very small, and is built without medium-model support. This was done because I've just barely got ELKS running in blink16. Since the ELKS kernel didn't have MM support compiled in, it actually thinks the executable is a shell script, which you can see eventually displays "Syntax error: ')' expected. That's the ash shell trying to run ./elksmoria!

I hope this all makes sense. If not, please ask more questions.

Thank you!

**Comment by tyama501** [comment]
Hello @ghaerr ,

I understand the blink16 does not support medium model yet and if I do blink16 elksmori then it says "not supported medium model" .

Also the reason why I am using the system mode is that I need to put morifiles in root directory.
Is there way to put files when direct mode?

The banner is working on the direct mode.

Thank you.

**Comment by ghaerr** [comment]
> Also the reason why I am using the system mode is that I need to put morifiles in root directory.
> Is there way to put files when direct mode?

Currently, the blink16 system call translation just translates `open` and doesn't change the pathname - this would mean you'd need the files in the root directory of your host system, which is not a good idea. I will have to think more about how to best handle this, and have made a note of it. For the time being, elksmoria won't run anyways since the medium model loader isn't done yet. Sounds like we still have lots of work to do to get blink16 usable, it just isn't ready yet for larger programs nor programs that use fixed pathnames. For that reason, I would suggest testing with small programs, as there is so much left to get working.

**Comment by charlysan** [comment]
Sorry if this is not related to the main topic, but is there any alternative for having an assembler and debugger within ELKS? So that we could generate binaries from x86 asm code?

**Comment by ghaerr** [comment]
>  is there any alternative for having an assembler and debugger within ELKS? So that we could generate binaries from x86 asm code?

Do you mean operating from ELKS itself, or just an alternative to assemble ASM programs from your host to run on ELKS? 

We don't have much that can run on ELKS itself at this time (see #1443 for C compiler discussions). But `nasm` or another assembler could be used to generate binaries for ELKS. We would need to generate the output ELKS-compatible a.out header from an ASM file as well for it to be loaded. Of course, our current toolchain will assemble source into a.out ELKS binaries, using AT&T format input.

**Comment by charlysan** [comment]
Thanks @ghaerr. I meant operating from ELKS itself. I saw the "C compiler" discussion and I understand the difficulties, but I thought that maybe for the case of an assembler operating from ELKS it could be easier than the C compiler (?).

**Comment by cocus** [comment]
Hi all, this looks so cool! In fact, it's so cool, that I dusted off my 8018x and flashed it with today's tip of master.
The `text-ui` came pre-enabled by default, but that doesn't fit on my ROMFS (I think it's because the actual kernel is placed right after the ROMFS and I should move it a little bit); but I was able to copy the binaries to an SD and run them.
I was able to run the `fm` fine, although really slow (a little bit faster over ethernet for sure!), but `cons` doesn't do anything interesting, except writing colored-characters all over the screen (either via serial or telnet, with putty or picocom):
![image](https://github.com/jbruchon/elks/assets/523726/a88b2bc2-6489-4060-9638-360c9d5e8ec1)
I would like to run the memopad or something cool, and feel the "d-flat" experience (which I was looking for like 4 months ago, never knew the name of this "windowing" scheme).
Do I have to do something special in order for this to work?
Thank you!

**Comment by ghaerr** [comment]
Hello @cocus,

Nice to hear from you, how have you been doing?

> but cons doesn't do anything interesting, except writing colored-characters all over the screen 

The reason `cons` doesn't work on your 8018x is that it reads the IBM PC BIOS display screen area at B800:0, which appears to be either random memory or non-existent on your system. It would be possible to hack the direct console with some work, which could then run headless and then use `cons` to display the pseudo-console, but that's probably not worth the effort, and the keyboard would have to be dealt with separately.

> I would like to run the memopad or something cool, and feel the "d-flat" experience

Because of its size, D-Flat and `memopad` are built in a seperate repository, for both UNIX and ELKS. You can take a look at it over [here](https://github.com/ghaerr/dflat) and build the ELKS version using `make -f Makefile.elks`.

> The text-ui came pre-enabled by default

What are you referring to here?

Thank you!

**Comment by cocus** [comment]
> Nice to hear from you, how have you been doing?
> 
Pretty good! Thank you! How are you doing? I see you were really busy with elks!


> The reason `cons` doesn't work on your 8018x is that it reads the IBM PC BIOS display screen area at B800:0, which appears to be either random memory or non-existent on your system. It would be possible to hack the direct console with some work, which could then run headless and then use `cons` to display the pseudo-console, but that's probably not worth the effort, and the keyboard would have to be dealt with separately.
>

Ah! I thought `cons` was required for memopad to run. I was mistaken. Yes, I agree, completely useless to do this on the headless targets



> What are you referring to here?
>

Oh, if I copy the 8018x.config as my main .config and run the menuconfig, I see that the `text-ui` was enabled by default on the userland. 




> Because of its size, D-Flat and `memopad` are built in a seperate repository, for both UNIX and ELKS. You can take a look at it over [here](https://github.com/ghaerr/dflat) and build the ELKS version using `make -f Makefile.elks`.
>

Ah! cool. I just built it (I had to change the path to my elks TOPDIR under the Makefile.elks) but the binary doesn't run on my board, I always get a 
```
# ./memopad
0@�@^: not found
./memopad: 2: Syntax error: ")" unexpected
```
the file seems to be okay. I copied it to the SD card because it won't fit on the rootfs romfs.

**Comment by ghaerr** [comment]
> the binary doesn't run on my board,

Oh, `memopad` is built in medium model, since its code size > 64k. That means the kernel needs to have the a.out medium model extension enabled: set `CONFIG_EXEC_MMODEL=y` and rebuild the kernel. (Sorry if that's a pain to reflash!)

> I see that the text-ui was enabled by default on the userland.

I'm sorry I'm being a little dense here, but I still don't know what `text-ui` is referring to, I don't recall having ever seen that term used within ELKS.

**Comment by cocus** [comment]
> Oh, `memopad` is built in medium model, since its code size > 64k. That means the kernel needs to have the a.out medium model extension enabled: set `CONFIG_EXEC_MMODEL=y` and rebuild the kernel. (Sorry if that's a pain to reflash!)

Okay, I'll test this. What are the downsides of enabling this? can I do it and keep it as the default?



> I'm sorry I'm being a little dense here, but I still don't know what `text-ui` is referring to, I don't recall having ever seen that term used within ELKS.

It's the name set on the menuconfig for `tui`:
![image](https://github.com/jbruchon/elks/assets/523726/2dd85acb-9337-4005-b315-6af6153def6d)

**Comment by ghaerr** [comment]
> What are the downsides of enabling this? can I do it and keep it as the default?

There are no downsides. Its only a config option because of the extra space taken up in the kernel, which is small. Perhaps it should be on by default. (Though not many programs are medium model at this point, I think memopad is the only one :)

> It's the name set on the menuconfig for tui:

Geez, thanks. Yes, I named that and promptly forgot what I called it. Are you thinking it should be off by default for 8018x?

**Comment by cocus** [comment]
> There are no downsides. Its only a config option because of the extra space taken up in the kernel, which is small. Perhaps it should be on by default. (Though not many programs are medium model at this point, I think memopad is the only one :)


Well, I enabled it, but it still doesn't run. Could it be because I don't have enough memory? meminfo reports 360kb free.
```
# ./memopad 
# 
# ./memopad 
# 
```



> Geez, thanks. Yes, I named that and promptly forgot what I called it. Are you thinking it should be off by default for 8018x?

Well, I think it might be a good idea to turn it off by default, just for the sake of reclaiming some bytes on the romfs

**Comment by ghaerr** [comment]
Does trying to run memopad act like a shell script command failure, or does the shell return an exec error? If the former, the a.out magic number is not correct, which means either no kernel support or bad executable. Post your memopad and I’ll run it when I get home.

Push text ui off on your open PR and we’ll commit it there.

On May 20, 2023, at 4:46 PM, Santiago Hormazabal ***@***.***> wrote:

﻿

There are no downsides. Its only a config option because of the extra space taken up in the kernel, which is small. Perhaps it should be on by default. (Though not many programs are medium model at this point, I think memopad is the only one :)

Well, I enabled it, but it still doesn't run. Could it be because I don't have enough memory? meminfo reports 360kb free.

Geez, thanks. Yes, I named that and promptly forgot what I called it. Are you thinking it should be off by default for 8018x?

Well, I think it might be a good idea to turn it off by default, just for the sake of reclaiming some bytes on the romfs

**Comment by cocus** [comment]
> Does trying to run memopad act like a shell script command failure, or does the shell return an exec error? If the former, the a.out magic number is not correct, which means either no kernel support or bad executable. Post your memopad and I’ll run it when I get home. Push text ui off on your open PR and we’ll commit it there.

I just pushed the config with the tui disabled, but I need to enable the medium model config. I'll do that in a second.

When I run it, no error is shown:
```
# cd /mnt
# ./memopad
# echo $?
0
```

**Comment by ghaerr** [comment]
Ok it is no longer the shell script syntax error, so the kernel is or should be loading the executable. It could be failing due to no memory, but what you are showing is the exit code, not the exec errno.

It could also be trying to write to IBM PC console memory instead of emitting ANSI sequences. I think there may be some ifdef ELKS in the source code that forces that. I will have a look in a couple hours.


On May 20, 2023, at 4:56 PM, Santiago Hormazabal ***@***.***> wrote:

﻿

Does trying to run memopad act like a shell script command failure, or does the shell return an exec error? If the former, the a.out magic number is not correct, which means either no kernel support or bad executable. Post your memopad and I’ll run it when I get home. Push text ui off on your open PR and we’ll commit it there.

I just pushed the config with the tui disabled, but I need to enable the medium model config. I'll do that in a second.

When I run it, no error is shown:

# cd /mnt
# ./memopad
# echo $?
0

**Comment by ghaerr** [comment]
Hello @cocus,

> I enabled it, but it still doesn't run. Could it be because I don't have enough memory? meminfo reports 360kb free.

It looks like `memopad` needs an 84k code segment and is requesting (and getting) a 65k data segment (see second screenshot for these), which matches the 147k or so shown by `ps`:
![DF ps 1](https://github.com/jbruchon/elks/assets/11985637/6d4c396a-3280-4e51-a819-342e24880bb8)

<img width="832" alt="DF heap" src="https://github.com/jbruchon/elks/assets/11985637/c497205b-032f-4a22-9692-daf6e4fc96d2">



Can you post a `meminfo` of your system after boot and login? That will show what the free memory looks like.

I'm also trying to track down whether `sh` or `sash` will display an error when errno=ENOMEM is returned from `exec`, which should be happening, but likely is not.

(I'd like to add an option to `meminfo` that would make it easier to show just application used/available memory, rather than all the other stuff... its getting kind of long).

Thank you!

**Comment by cocus** [comment]
> Can you post a `meminfo` of your system after boot and login? That will show what the free memory looks like.

```
# meminfo
  HEAP   TYPE  SIZE    SEG   TYPE    SIZE  CNT
  ab7e   SEG     16   1080   CSEG    7056    1
  ab9a   INT      6
  abac   INT      6
  abbe   INT      6
  abd0   INT      6
  abe2   BUFH   216
  acc6   TTY     80
  ad22   TTY     80
  ad7e   SEG     16   1239   DSEG    5808    1
  ad9a   SEG     16   13a4   free   24864    0
  adb6   SEG     16   25e6   DSEG   24864    1
  add2   SEG     16   2bf8   DSEG    2352    1
  adee   SEG     16   19b6   CSEG    3632    1
  ae0a   SEG     16   1a99   free    2320    0
  ae26   SEG     16   1b2a   CSEG   43968    1
  ae42   SEG     16   2c8b   free  341840    0
  ae5e   free 20898
  Heap/free   21646/20898 Total mem  456704
  Memory usage  446KB total,   86KB used,  360KB free
# /mnt/memopad
# 
```

**Comment by ghaerr** [comment]
Thanks @cocus. You're running this from the external headless console to Putty, right? Can you do a `ps` and tell me what the TTY number is you're running on ? `memopad` does try to do something slightly different, by not translating all the CP 437 characters to ANSI Unicode sequences, when running on /dev/tty1 but that shouldn't affecting this. I can't yet determine whether its running, crashing, or the kernel is rejecting it with ENOMEM and /bin/sh isn't saying so. The meminfo above shows plenty of room for the executable. If you'd like to post `memopad` here, I'll try running it on QEMU (or you can try that also - the D-Flat Makefile.elks copies it to ELKS /root, I happen to notice).

**Comment by cocus** [comment]
> Thanks @cocus. You're running this from the external headless console to Putty, right? Can you do a `ps` and tell me what the TTY number is you're running on ?
Yes, this is through the headless console, The same happens when I run it through telnet, although with less available RAM.
I'm using picocom, but the same happens with Putty.
```
# ps
  PID   GRP  TTY USER STAT CPU  HEAP  FREE   SIZE COMMAND
    1     0      root    S   0  3072  2016  12864 /bin/init 
    6     6    1 root    S   1  1150 13234  68832 -/bin/sh 
   11     6    1 root    R   0  1024  1176  11264 ps
```

> `memopad` does try to do something slightly different, by not translating all the CP 437 characters to ANSI Unicode sequences, when running on /dev/tty1 but that shouldn't affecting this. I can't yet determine whether its running, crashing, or the kernel is rejecting it with ENOMEM and /bin/sh isn't saying so. The meminfo above shows plenty of room for the executable. If you'd like to post `memopad` here, I'll try running it on QEMU (or you can try that also - the D-Flat Makefile.elks copies it to ELKS /root, I happen to notice).

Here it is:
https://www.dropbox.com/s/fhb81xqzvkhk2oe/memopad?dl=0

**Comment by ghaerr** [comment]
Thanks, your version runs fine on QEMU, and is the same as mine.

I see that your system is running headless on /dev/tty1. That means `memopad` (the D-Flat library) is thinking the system is running on the ELKS console. 

Try rebuilding `memopad` with the following patch and rerunning it:
```
diff --git a/tty.c b/tty.c
index a519134..030a84f 100644
--- a/tty.c
+++ b/tty.c
@@ -110,6 +110,7 @@ int tty_init(enum ttyflags f)
         once = 1;
         memcpy(&t, &oldterm, sizeof(t));
         iselksconsole = tty_iselksconsole(1);
+        iselksconsole = 0;
     }
     tty_enable_unikey();
     if (flags & FullBuffer)
```

**Comment by cocus** [comment]
I changed that, re build it, and tested it again, but this time we're back with the old errors:
```
# /mnt/memopad
0@�@^: not found
/mnt/memopad: 2: Syntax error: ")" unexpected
```
I'm not sure why, the medium model thing is enabled on my kernel.

**Comment by ghaerr** [comment]
Well, something is goofed up for sure. To determine if you've got a valid executable you can run `chmem`:
```
ELKS 0.6.0

login: root
# chmem ./memopad 
 TEXT  FTEXT   DATA    BSS   HEAP  STACK  TOTDATA  TOTAL
42736  41520  24128   6944  65535      0    65520 149776 ./memopad
# 
```
If it shows as valid (with approx 84k text and heap set to 65K which means max heap), then it means that your kernel isn't running CONFIG_EXEC_MMODEL. Otherwise, we've got a build issue with `make elks` over at D-Flat.

**Comment by cocus** [comment]
> If it shows as valid (with approx 84k text and heap set to 65K which means max heap), then it means that your kernel isn't running CONFIG_EXEC_MMODEL. Otherwise, we've got a build issue with `make elks` over at D-Flat.

i've got the same output as you, so the MMODEL is not set then? let me clean everything on the elks topdir and re-build the images

```
# /mnt/chmem /mnt/memopad 
 TEXT  FTEXT   DATA    BSS   HEAP  STACK  TOTDATA  TOTAL
42736  41520  24128   6944  65535      0    65520 149776 /mnt/memopad
# /mnt/memopad
0@�@^: not found
/mnt/memopad: 2: Syntax error: ")" unexpected
```

**Comment by ghaerr** [comment]
Sounds like `CONFIG_EXEC_MMODEL` is not set. However, your `chmem` should be slightly different than mine with the `iselksconsole = 0;` patch included.

**Comment by cocus** [comment]
![image](https://github.com/jbruchon/elks/assets/523726/633c0f4a-ea61-4e99-b94f-abb8e13890f8)

hey hey! this is so painfully slow over serial, that I might die! but it works!!!! how cool is this!

I see the uart TX led constantly on, even while not doing absolutely anything on the terminal... it might be expected I guess

![image](https://github.com/jbruchon/elks/assets/523726/f6e066f2-6f96-4145-a13b-9e22ba24ac31)

**Comment by ghaerr** [comment]
> this is so painfully slow over serial, that I might die! but it works!!!! how cool is this!

Well it does look cool, even it it's hellishly slow?! Kind of neat that your ROM based 8018x is computing all this :)

What baud rate are you running? I'm not yet certain whether the bottleneck is just the 80c188eb calculating D-Flat stuff, or perhaps a dumb virtual screen to ANSI sequence conversion routine I wrote. I haven't actually ever run this over actual serial, only emulated on a fast laptop over pseudo terminals!

> I see the uart TX led constantly on, even while not doing absolutely anything on the terminal...

I just happened to notice something similar when coming up with the `iselksconsole` patch. I'll investigate that some more and also try to come up with a mechanism other than the current patch for allowing D-Flat to run on ELKS console and serial.
It is also possible the continuous transmitting could be part of the speed problem you're seeing.

**Comment by cocus** [comment]
> What baud rate are you running? I'm not yet certain whether the bottleneck is just the 80c188eb calculating D-Flat stuff, or perhaps a dumb virtual screen to ANSI sequence conversion routine I wrote. I haven't actually ever run this over actual serial, only emulated on a fast laptop over pseudo terminals!

I'd say it's both the CPU and the serial. It's running at 9600 by default (since the headless driver configures it by default, even if the actual driver supports faster baudrates). I've tried 38400 bps before, and it works, but the driver always initializes it at 9600 :(.

Running it thru telnet is "somewhat" faster, but still slow, so you can get an idea that the problem is the CPU more or less. Even if it runs at 16MHz.




> I just happened to notice something similar when coming up with the `iselksconsole` patch. I'll investigate that some more and also try to come up with a mechanism other than the current patch for allowing D-Flat to run on ELKS console and serial.
> It is also possible the continuous transmitting could be part of the speed problem you're seeing.

Yeah, no problems, just seeing this is amazing. Thank you for your efforts!

**Comment by ghaerr** [comment]
> I've tried 38400 bps before, and it works, but the driver always initializes it at 9600 :(.

You're welcome to submit a patch for CONFIG_ARCH_8018X to default serial baud rate to 38400. I'd do it for IBM PC except I think others may guess/expect 9600.

I'd be interested to know more accurately how much the baud rate affects the results here. Perhaps even running 57600 would help. If the speedup is significant, its not CPU bound. Another idea would be to implement interrupt-driven transmit, instead of busying looping for output. And `memopad` shouldn't be looping or continuously outputting byte streams, I'll look into that on QEMU.

Now you need a mouse (somehow). Is that possible on that system (there's only one serial port, right)?

**Comment by cocus** [comment]
> You're welcome to submit a patch for CONFIG_ARCH_8018X to default serial baud rate to 38400. I'd do it for IBM PC except I think others may guess/expect 9600.
> 
> I'd be interested to know more accurately how much the baud rate affects the results here. Perhaps even running 57600 would help. If the speedup is significant, its not CPU bound. Another idea would be to implement interrupt-driven transmit, instead of busying looping for output. And `memopad` shouldn't be looping or continuously outputting byte streams, I'll look into that on QEMU.

It'd be nice to know how to force the default speed to 38400, haven't figured how to do that myself.



> Now you need a mouse (somehow). Is that possible on that system (there's only one serial port, right)?

It has two serial ports, but the secondary one was never enabled, nor any drivers added to it. However, using the main serial port and a console emulator, I can use the mouse without any issues! It works fine! I was even able to drag and drop a window, slowly as hell, but they moved!

**Comment by ghaerr** [comment]
> It'd be nice to know how to force the default speed to 38400, haven't figured how to do that myself.

The 8018x port is using a special console implemented in elks/arch/i86/drivers/char/console-serial-8018x.c. In that file, the `console_init` function calls `update_port` which references the associated`struct tty` found in ntty.c. In ntty.c the `struct termios def_vals` is initialized with `B9600 | CS8` which could be ifdefed CONFIG_ARCH_8018X and set to `B38400 | CS8` (or perhaps set a DEFAULT_BAUD earlier in the file with an ifdef there).

[EDIT: Try 57600 or more, it would be interesting to see what the highest value the 8018x might support. At some point too high a baud rate may case loss of data from the mouse stream... ].

> using the main serial port and a console emulator, I can use the mouse without any issues!

You mean your terminal emulator allows using a mouse, which is interpreted and then sent out to the serial input stream by D-Flat and moves the mouse? That's pretty cool, since I initially designed D-Flat to take mouse movements from the macOS Terminal, but didn't realize this could be used with a serial-based terminal emulator that interfaced with a mouse and does the same thing! Very nice!

**Comment by cocus** [comment]
> The 8018x port is using a special console implemented in elks/arch/i86/drivers/char/console-serial-8018x.c. In that file, the `console_init` function calls `update_port` which references the associated`struct tty` found in ntty.c. In ntty.c the `struct termios def_vals` is initialized with `B9600 | CS8` which could be ifdefed CONFIG_ARCH_8018X and set to `B38400 | CS8` (or perhaps set a DEFAULT_BAUD earlier in the file with an ifdef there).

Aha, I'll try this! Good idea.



> [EDIT: Try 57600 or more, it would be interesting to see what the highest value the 8018x might support. At some point too high a baud rate may case loss of data from the mouse stream... ].

I'd love to use even more than that, but the problem is the FCPU that I have (which is set by the external crystal). The baudrate generator is really off when going above 38400. Usually my FTDI USB-serial doesn't even read anything, and I've seen this on the logic analyzer that the timings are really off, as the theory said.

It's funny because some other user with a different clock could easily use the higher baudrates without issues!



> You mean your terminal emulator allows using a mouse, which is interpreted and then sent out to the serial input stream by D-Flat and moves the mouse? That's pretty cool, since I initially designed D-Flat to take mouse movements from the macOS Terminal, but didn't realize this could be used with a serial-based terminal emulator that interfaced with a mouse and does the same thing! Very nice!

Yep, I did exactly that. It worked on my `telnet` and `picocom`, didn't test PuTTY but it usually works as well.

**Comment by tyama501** [comment]
> (I'd like to add an option to meminfo that would make it easier to show just application used/available memory, rather than all the other stuff... its getting kind of long).

I think so too.
It would be nice if meminfo can show names of applications like FreeDOS mem.
Is it hard to do that? 

FreeDOS mem 
<img width="437" alt="mem_freedos" src="https://github.com/jbruchon/elks/assets/61556504/f69d9588-1190-4c51-878a-2e09d2d1159f">

**Comment by ghaerr** [comment]
> It would be nice if meminfo can show names of applications like FreeDOS mem.
> Is it hard to do that?

It isn't that straightforward, but certainly could be done. I've been wanting the same feature as well for a while. Let me take a look at how we might get this working :)
